### ISSUE TITLE
Document when the new version of a remote function is used when the definition changes.

### ISSUE THREAD

**Issue by robertnishihara** [ray-project/ray#487]
We should document precisely what happens when a remote function is redefined (this should go in the troubleshooting documentation). In particular, does the new version get run or the old version? There are some subtleties.

**Case 1: `f` redefined.**

For example, if we define a function and then redefine it, the new version should run.

```python
@ray.remote
def f():
  return 1

@ray.remote
def f():
  return 2

ray.get(f.remote())  # This should be 2.
```

**Case 2: `f` in a file and reimported.**

What if the definition of `f` is in a file `temp.py`? Suppose we `import temp`, then change the definition of `f` in the file, then re-import `temp`.

In that case, the second import will be a no-op and the first definition will be used.

**Case 3: `f` in a file and reloaded.**

Again the definition of `f` is in a file `temp.py`. Suppose we `import temp`, then change the definition of `f` in the file, then reload `temp`.

In that case, the reload causes the new definition to get re-executed and exported to the other machines, so the new definition should be used.

**Case 4: helper  function in a file.**

This is a bit more subtle. Suppose that `f` uses a helper function defined in a file. Then when `f` is defined, its definition is shipped to all of the workers, and unpickled. When it is unpickled, the file with the helper function is imported. Now suppose that the helper file is changed. If we redefine `f`, then the definition of `f` is pickled again and shipped to the workers. When it is unpickled on a worker, the worker imports the helper file, but since the helper file has been imported already, this import is a no-op, so the new definition of the helper function **is not used**. Reloading on the driver doesn't help because the reload needs to happen on the worker.

**Comment by robertnishihara** [comment]
Consider the following case.

You have `helper.py` with

```python
def helper_function():
  return 1
```

And you define a remote function as follows.

```python
import helper

@ray.remote
def f():
  return helper.helper_function()
```

Then suppose you change the definition of `helper_function` in the file to return `2`. To avoid having to stop and restart Ray, the following workaround could help. Redfine `f` as follows.

```python
# Note that in Python 3 you need to do `from importlib import reload`

@ray.remote
def f():
  reload(helper)
  return helper.helper_function()
```

That should force the reload to happen on the workers when it is needed.

@hmania

**Comment by robertnishihara** [comment]
Addressed by #640.
